janusgraph-core/src/main/java/org/janusgraph/graphdb/util/MultiIterator.java

```diff
@@ -0,0 +1,59 @@
+// Copyright 2020 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.util;
+
+import org.apache.tinkerpop.gremlin.structure.util.CloseableIterator;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+
+public class MultiIterator<E> extends CloseableAbstractIterator<E> {
+
+    private Iterator<CloseableIterator<E>> listIterator;
+    private CloseableIterator<E> currentIterator;
+
+    public MultiIterator(final List<CloseableIterator<E>> iterators) {
+        Objects.requireNonNull(iterators);
+        listIterator = iterators.iterator();
```

Comment:
I've seen you used `IteratorUtils.concat(iterators);`. Will it be easier to manage `MultiIterator` with `IteratorUtils.concat(iterators);` or did you create it for some other purposes (like better performance, less overhead or something like that)?

Reply:
In fact that is the reason why `MultiIterator` was created at the first place:

```java
public static <T> CloseableIterator<T> concat(final List<CloseableIterator<T>> iterators) {
    return new MultiIterator<T>(iterators);
}
```

Reply:
Yap, I missed that. Now I see the purpose of this class. I'm just thinking if we can somehow reuse already existent general functionality instead of introducing our own implementation and then support it.
I know we have #2033 but I don't see it to be resolved anytime soon. Thus, I think, in case we don't have better alternatives for some functionality, it's okay to use that functionality from Guava.
For example, Guava already has this `concat` functionality in `Iterators`. So, basically you would be able to do something like this:
```
public static <T> CloseableIterator<T> concat(final List<CloseableIterator<T>> iterators) {
    return Iterators.concat(iterators.iterator());
}
```
Instead of supporting this `MultiIterator`, but I see that `MultiIterator` not just concat iterators as `Iterators.concat` does, but also closes iterators which don't have any value anymore.
So, I was thinking, if it is right. I'm not sure yet. Maybe it is a correct implementation but for me it's a little bit confusing.
My confusion is that it's not clear, when exactly we trigger `close` for `CloseableIterator`. Looking in this PR I see that some Iterators automatically trigger `close` after the last element they read but some Iterators don't trigger this logic and expect external logic to trigger `close` for them (like `LimitAdjustingIterator`).
I need to look at this PR a little bit more deeply. I have a small feeling that we can remove our `concat` implementation and if each Iterator which implements `CloseableIterator` would close itself automatically, then we don't need to pass `CloseableIterator` iterator everywhere and we don't need to wrap `Iterator` as `CloseableIterator` when it doesn't have anything to close. That said, I could be completely wrong, I just didn't check the full PR yet.